// @create-markdown/preview/dist/themes/css-strings.js
// src/themes/css-strings.ts
var githubDark = `/**
 * @create-markdown/preview - GitHub Dark Theme
 * Dark theme inspired by GitHub's dark mode markdown rendering
 */

.cm-preview {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Noto Sans', Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 1.6;
  color: #e6edf3;
  background-color: #0d1117;
  max-width: 100%;
  overflow-wrap: break-word;
}

/* Headings */
.cm-heading {
  margin-top: 24px;
  margin-bottom: 16px;
  font-weight: 600;
  line-height: 1.25;
}

.cm-h1 {
  font-size: 2em;
  padding-bottom: 0.3em;
  border-bottom: 1px solid #30363d;
}

.cm-h2 {
  font-size: 1.5em;
  padding-bottom: 0.3em;
  border-bottom: 1px solid #30363d;
}

.cm-h3 { font-size: 1.25em; }
.cm-h4 { font-size: 1em; }
.cm-h5 { font-size: 0.875em; }
.cm-h6 { font-size: 0.85em; color: #8b949e; }

/* Paragraphs */
.cm-paragraph {
  margin-top: 0;
  margin-bottom: 16px;
}

/* Lists */
.cm-bullet-list,
.cm-numbered-list {
  margin-top: 0;
  margin-bottom: 16px;
  padding-left: 2em;
}

.cm-bullet-list li,
.cm-numbered-list li {
  margin-bottom: 4px;
}

/* Checklist */
.cm-checklist-item {
  display: flex;
  align-items: flex-start;
  gap: 0.5rem;
  margin-bottom: 4px;
}

.cm-checklist-item input[type="checkbox"] {
  margin-top: 0.25rem;
}

.cm-checked {
  text-decoration: line-through;
  color: #8b949e;
}

/* Code */
.cm-code-block {
  background-color: #161b22;
  border-radius: 6px;
  padding: 16px;
  overflow-x: auto;
  margin-bottom: 16px;
  font-size: 0.875em;
}

.cm-code {
  font-family: ui-monospace, SFMono-Regular, 'SF Mono', Menlo, Consolas, 'Liberation Mono', monospace;
  background: transparent;
  padding: 0;
  white-space: pre;
}

code {
  font-family: ui-monospace, SFMono-Regular, 'SF Mono', Menlo, Consolas, 'Liberation Mono', monospace;
  font-size: 0.875em;
  background-color: rgba(110, 118, 129, 0.4);
  padding: 0.2em 0.4em;
  border-radius: 6px;
}

/* Blockquote */
.cm-blockquote {
  margin: 0 0 16px 0;
  padding: 0 1em;
  color: #8b949e;
  border-left: 0.25em solid #30363d;
}

/* Divider */
.cm-divider {
  height: 0.25em;
  padding: 0;
  margin: 24px 0;
  background-color: #30363d;
  border: 0;
}

/* Table */
.cm-table {
  display: block;
  width: max-content;
  max-width: 100%;
  overflow: auto;
  border-spacing: 0;
  border-collapse: collapse;
  margin-bottom: 16px;
}

.cm-table th,
.cm-table td {
  padding: 6px 13px;
  border: 1px solid #30363d;
}

.cm-table th {
  font-weight: 600;
  background-color: #161b22;
}

.cm-table tr:nth-child(2n) {
  background-color: #161b22;
}

/* Image */
.cm-image {
  margin: 0 0 16px 0;
}

.cm-image img {
  max-width: 100%;
  height: auto;
  border-radius: 6px;
}

.cm-image figcaption {
  font-size: 0.875em;
  color: #8b949e;
  margin-top: 8px;
  text-align: center;
}

/* Callout */
.cm-callout {
  padding: 16px;
  margin-bottom: 16px;
  border-radius: 6px;
  border-left: 4px solid;
}

.cm-callout-title {
  display: block;
  text-transform: capitalize;
  margin-bottom: 8px;
}

.cm-callout-info {
  border-left-color: #58a6ff;
  background-color: rgba(56, 139, 253, 0.15);
}

.cm-callout-warning {
  border-left-color: #d29922;
  background-color: rgba(187, 128, 9, 0.15);
}

.cm-callout-tip {
  border-left-color: #3fb950;
  background-color: rgba(46, 160, 67, 0.15);
}

.cm-callout-danger {
  border-left-color: #f85149;
  background-color: rgba(248, 81, 73, 0.15);
}

.cm-callout-note {
  border-left-color: #8b949e;
  background-color: rgba(110, 118, 129, 0.15);
}

/* Links */
.cm-preview a {
  color: #58a6ff;
  text-decoration: none;
}

.cm-preview a:hover {
  text-decoration: underline;
}

/* Inline styles */
.cm-preview mark {
  background-color: rgba(187, 128, 9, 0.4);
  padding: 0.1em 0.2em;
  border-radius: 3px;
}

.cm-preview del {
  color: #8b949e;
}
`;
var github = `/**
 * @create-markdown/preview - GitHub Theme
 * Light theme inspired by GitHub's markdown rendering
 */

.cm-preview {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Noto Sans', Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 1.6;
  color: #1f2328;
  background-color: #ffffff;
  max-width: 100%;
  overflow-wrap: break-word;
}

/* Headings */
.cm-heading {
  margin-top: 24px;
  margin-bottom: 16px;
  font-weight: 600;
  line-height: 1.25;
}

.cm-h1 {
  font-size: 2em;
  padding-bottom: 0.3em;
  border-bottom: 1px solid #d1d9e0;
}

.cm-h2 {
  font-size: 1.5em;
  padding-bottom: 0.3em;
  border-bottom: 1px solid #d1d9e0;
}

.cm-h3 { font-size: 1.25em; }
.cm-h4 { font-size: 1em; }
.cm-h5 { font-size: 0.875em; }
.cm-h6 { font-size: 0.85em; color: #656d76; }

/* Paragraphs */
.cm-paragraph {
  margin-top: 0;
  margin-bottom: 16px;
}

/* Lists */
.cm-bullet-list,
.cm-numbered-list {
  margin-top: 0;
  margin-bottom: 16px;
  padding-left: 2em;
}

.cm-bullet-list li,
.cm-numbered-list li {
  margin-bottom: 4px;
}

/* Checklist */
.cm-checklist-item {
  display: flex;
  align-items: flex-start;
  gap: 0.5rem;
  margin-bottom: 4px;
}

.cm-checklist-item input[type="checkbox"] {
  margin-top: 0.25rem;
}

.cm-checked {
  text-decoration: line-through;
  color: #656d76;
}

/* Code */
.cm-code-block {
  background-color: #f6f8fa;
  border-radius: 6px;
  padding: 16px;
  overflow-x: auto;
  margin-bottom: 16px;
  font-size: 0.875em;
}

.cm-code {
  font-family: ui-monospace, SFMono-Regular, 'SF Mono', Menlo, Consolas, 'Liberation Mono', monospace;
  background: transparent;
  padding: 0;
  white-space: pre;
}

code {
  font-family: ui-monospace, SFMono-Regular, 'SF Mono', Menlo, Consolas, 'Liberation Mono', monospace;
  font-size: 0.875em;
  background-color: rgba(175, 184, 193, 0.2);
  padding: 0.2em 0.4em;
  border-radius: 6px;
}

/* Blockquote */
.cm-blockquote {
  margin: 0 0 16px 0;
  padding: 0 1em;
  color: #656d76;
  border-left: 0.25em solid #d1d9e0;
}

/* Divider */
.cm-divider {
  height: 0.25em;
  padding: 0;
  margin: 24px 0;
  background-color: #d1d9e0;
  border: 0;
}

/* Table */
.cm-table {
  display: block;
  width: max-content;
  max-width: 100%;
  overflow: auto;
  border-spacing: 0;
  border-collapse: collapse;
  margin-bottom: 16px;
}

.cm-table th,
.cm-table td {
  padding: 6px 13px;
  border: 1px solid #d1d9e0;
}

.cm-table th {
  font-weight: 600;
  background-color: #f6f8fa;
}

.cm-table tr:nth-child(2n) {
  background-color: #f6f8fa;
}

/* Image */
.cm-image {
  margin: 0 0 16px 0;
}

.cm-image img {
  max-width: 100%;
  height: auto;
  border-radius: 6px;
}

.cm-image figcaption {
  font-size: 0.875em;
  color: #656d76;
  margin-top: 8px;
  text-align: center;
}

/* Callout */
.cm-callout {
  padding: 16px;
  margin-bottom: 16px;
  border-radius: 6px;
  border-left: 4px solid;
}

.cm-callout-title {
  display: block;
  text-transform: capitalize;
  margin-bottom: 8px;
}

.cm-callout-info {
  border-left-color: #0969da;
  background-color: #ddf4ff;
}

.cm-callout-warning {
  border-left-color: #d4a72c;
  background-color: #fff8c5;
}

.cm-callout-tip {
  border-left-color: #1a7f37;
  background-color: #dafbe1;
}

.cm-callout-danger {
  border-left-color: #cf222e;
  background-color: #ffebe9;
}

.cm-callout-note {
  border-left-color: #656d76;
  background-color: #f6f8fa;
}

/* Links */
.cm-preview a {
  color: #0969da;
  text-decoration: none;
}

.cm-preview a:hover {
  text-decoration: underline;
}

/* Inline styles */
.cm-preview mark {
  background-color: #fff8c5;
  padding: 0.1em 0.2em;
  border-radius: 3px;
}

.cm-preview del {
  color: #656d76;
}
`;
var minimal = "/**\n * @create-markdown/preview - Minimal Theme\n * A clean, minimal theme with subtle styling\n */\n\n.cm-preview {\n  font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, sans-serif;\n  font-size: 16px;\n  line-height: 1.7;\n  color: #333;\n  max-width: 100%;\n  overflow-wrap: break-word;\n}\n\n/* Headings */\n.cm-heading {\n  margin-top: 1.5em;\n  margin-bottom: 0.5em;\n  font-weight: 600;\n  line-height: 1.3;\n}\n\n.cm-h1 { font-size: 1.75em; }\n.cm-h2 { font-size: 1.5em; }\n.cm-h3 { font-size: 1.25em; }\n.cm-h4 { font-size: 1.1em; }\n.cm-h5 { font-size: 1em; }\n.cm-h6 { font-size: 0.9em; color: #666; }\n\n/* Paragraphs */\n.cm-paragraph {\n  margin-top: 0;\n  margin-bottom: 1em;\n}\n\n/* Lists */\n.cm-bullet-list,\n.cm-numbered-list {\n  margin-top: 0;\n  margin-bottom: 1em;\n  padding-left: 1.5em;\n}\n\n.cm-bullet-list li,\n.cm-numbered-list li {\n  margin-bottom: 0.25em;\n}\n\n/* Checklist */\n.cm-checklist-item {\n  display: flex;\n  align-items: flex-start;\n  gap: 0.5rem;\n  margin-bottom: 0.25em;\n}\n\n.cm-checked {\n  text-decoration: line-through;\n  opacity: 0.6;\n}\n\n/* Code */\n.cm-code-block {\n  background-color: #f5f5f5;\n  border-radius: 4px;\n  padding: 1em;\n  overflow-x: auto;\n  margin-bottom: 1em;\n  font-size: 0.9em;\n}\n\n.cm-code {\n  font-family: 'SF Mono', Monaco, 'Cascadia Code', 'Roboto Mono', monospace;\n  background: transparent;\n  padding: 0;\n  white-space: pre;\n}\n\ncode {\n  font-family: 'SF Mono', Monaco, 'Cascadia Code', 'Roboto Mono', monospace;\n  font-size: 0.9em;\n  background-color: #f0f0f0;\n  padding: 0.15em 0.3em;\n  border-radius: 3px;\n}\n\n/* Blockquote */\n.cm-blockquote {\n  margin: 0 0 1em 0;\n  padding: 0 1em;\n  color: #666;\n  border-left: 3px solid #ddd;\n  font-style: italic;\n}\n\n/* Divider */\n.cm-divider {\n  height: 1px;\n  padding: 0;\n  margin: 2em 0;\n  background-color: #eee;\n  border: 0;\n}\n\n/* Table */\n.cm-table {\n  display: block;\n  width: max-content;\n  max-width: 100%;\n  overflow: auto;\n  border-spacing: 0;\n  border-collapse: collapse;\n  margin-bottom: 1em;\n}\n\n.cm-table th,\n.cm-table td {\n  padding: 0.5em 1em;\n  border: 1px solid #ddd;\n}\n\n.cm-table th {\n  font-weight: 600;\n  background-color: #f9f9f9;\n}\n\n/* Image */\n.cm-image {\n  margin: 0 0 1em 0;\n}\n\n.cm-image img {\n  max-width: 100%;\n  height: auto;\n}\n\n.cm-image figcaption {\n  font-size: 0.875em;\n  color: #666;\n  margin-top: 0.5em;\n  text-align: center;\n}\n\n/* Callout */\n.cm-callout {\n  padding: 1em;\n  margin-bottom: 1em;\n  border-radius: 4px;\n  border-left: 3px solid;\n  background-color: #f9f9f9;\n}\n\n.cm-callout-title {\n  display: block;\n  text-transform: capitalize;\n  font-weight: 600;\n  margin-bottom: 0.5em;\n}\n\n.cm-callout-info { border-left-color: #3b82f6; }\n.cm-callout-warning { border-left-color: #f59e0b; }\n.cm-callout-tip { border-left-color: #10b981; }\n.cm-callout-danger { border-left-color: #ef4444; }\n.cm-callout-note { border-left-color: #6b7280; }\n\n/* Links */\n.cm-preview a {\n  color: #0066cc;\n  text-decoration: none;\n}\n\n.cm-preview a:hover {\n  text-decoration: underline;\n}\n\n/* Inline styles */\n.cm-preview mark {\n  background-color: #fff3cd;\n  padding: 0.1em 0.2em;\n  border-radius: 2px;\n}\n\n.cm-preview del {\n  opacity: 0.6;\n}\n";
var system = `/**
 * @create-markdown/preview - System Theme
 * CSS custom property theme that integrates with any design system.
 * Override --cm-* variables, or fall back to common --text / --bg / --border tokens.
 */

.cm-preview {
  font-family: var(--cm-font, inherit);
  font-size: var(--cm-font-size, 1rem);
  line-height: var(--cm-line-height, 1.6);
  color: var(--cm-text, var(--text, inherit));
  background: var(--cm-bg, var(--bg, transparent));
  max-width: 100%;
  overflow-wrap: break-word;
}

/* Headings */
.cm-heading {
  margin-top: 1.5em;
  margin-bottom: 0.5em;
  font-weight: var(--cm-heading-weight, 600);
  line-height: 1.25;
}

.cm-h1 {
  font-size: 2em;
  padding-bottom: 0.3em;
  border-bottom: 1px solid var(--cm-border, var(--border, currentColor));
}

.cm-h2 {
  font-size: 1.5em;
  padding-bottom: 0.3em;
  border-bottom: 1px solid var(--cm-border, var(--border, currentColor));
}

.cm-h3 { font-size: 1.25em; }
.cm-h4 { font-size: 1em; }
.cm-h5 { font-size: 0.875em; }
.cm-h6 { font-size: 0.85em; color: var(--cm-muted, var(--muted, #656d76)); }

/* Paragraphs */
.cm-paragraph {
  margin-top: 0;
  margin-bottom: 1em;
}

/* Lists */
.cm-bullet-list,
.cm-numbered-list {
  margin-top: 0;
  margin-bottom: 1em;
  padding-left: 2em;
}

.cm-bullet-list li,
.cm-numbered-list li {
  margin-bottom: 0.25em;
}

/* Checklist */
.cm-checklist-item {
  display: flex;
  align-items: flex-start;
  gap: 0.5rem;
  margin-bottom: 0.25em;
}

.cm-checklist-item input[type="checkbox"] {
  margin-top: 0.25rem;
}

.cm-checked {
  text-decoration: line-through;
  color: var(--cm-muted, var(--muted, #656d76));
}

/* Code */
.cm-code-block {
  background: var(--cm-code-bg, var(--secondary, #f5f5f5));
  border: 1px solid var(--cm-border, var(--border, #ddd));
  border-radius: var(--cm-radius, 6px);
  padding: 1em;
  overflow-x: auto;
  margin-bottom: 1em;
  font-size: 0.875em;
}

.cm-code {
  font-family: var(--cm-mono, ui-monospace, SFMono-Regular, 'SF Mono', Menlo, Consolas, monospace);
  background: transparent;
  padding: 0;
  white-space: pre;
}

code {
  font-family: var(--cm-mono, ui-monospace, SFMono-Regular, 'SF Mono', Menlo, Consolas, monospace);
  font-size: 0.875em;
  background-color: var(--cm-inline-code-bg, var(--secondary, rgba(175, 184, 193, 0.2)));
  padding: 0.2em 0.4em;
  border-radius: var(--cm-radius, 6px);
}

/* Blockquote */
.cm-blockquote {
  margin: 0 0 1em 0;
  padding: 0 1em;
  color: var(--cm-muted, var(--muted, #656d76));
  border-left: 0.25em solid var(--cm-border, var(--border, #ddd));
}

/* Divider */
.cm-divider {
  height: 0.25em;
  padding: 0;
  margin: 1.5em 0;
  background-color: var(--cm-border, var(--border, #ddd));
  border: 0;
}

/* Table */
.cm-table {
  display: block;
  width: max-content;
  max-width: 100%;
  overflow: auto;
  border-spacing: 0;
  border-collapse: collapse;
  margin-bottom: 1em;
}

.cm-table th,
.cm-table td {
  padding: 6px 13px;
  border: 1px solid var(--cm-border, var(--border, #ddd));
}

.cm-table th {
  font-weight: 600;
  background-color: var(--cm-table-header-bg, var(--secondary, #f6f8fa));
}

.cm-table tr:nth-child(2n) {
  background-color: var(--cm-table-stripe-bg, var(--secondary, #f6f8fa));
}

/* Image */
.cm-image {
  margin: 0 0 1em 0;
}

.cm-image img {
  max-width: 100%;
  height: auto;
  border-radius: var(--cm-radius, 6px);
}

.cm-image figcaption {
  font-size: 0.875em;
  color: var(--cm-muted, var(--muted, #656d76));
  margin-top: 0.5em;
  text-align: center;
}

/* Callout */
.cm-callout {
  padding: 1em;
  margin-bottom: 1em;
  border-radius: var(--cm-radius, 6px);
  border-left: 4px solid;
  background-color: var(--cm-callout-bg, var(--secondary, #f6f8fa));
}

.cm-callout-title {
  display: block;
  text-transform: capitalize;
  font-weight: 600;
  margin-bottom: 0.5em;
}

.cm-callout-info { border-left-color: var(--cm-info, #0969da); }
.cm-callout-warning { border-left-color: var(--cm-warning, #d4a72c); }
.cm-callout-tip { border-left-color: var(--cm-success, #1a7f37); }
.cm-callout-danger { border-left-color: var(--cm-danger, #cf222e); }
.cm-callout-note { border-left-color: var(--cm-muted, var(--muted, #656d76)); }

/* Links */
.cm-preview a {
  color: var(--cm-link, var(--link, #0969da));
  text-decoration: none;
}

.cm-preview a:hover {
  text-decoration: underline;
}

/* Inline styles */
.cm-preview mark {
  background-color: var(--cm-highlight, #fff8c5);
  padding: 0.1em 0.2em;
  border-radius: 3px;
}

.cm-preview del {
  color: var(--cm-muted, var(--muted, #656d76));
}
`;
var themeCSS = {
  githubDark,
  github,
  minimal,
  system
};
export {
  github,
  githubDark,
  minimal,
  system,
  themeCSS
};
