// @create-markdown/preview/dist/web-component.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);

// src/web-component.ts
var web_component_exports = {};
__export(web_component_exports, {
  MarkdownPreviewElement: () => MarkdownPreviewElement,
  autoRegister: () => autoRegister,
  registerPreviewElement: () => registerPreviewElement
});
module.exports = __toCommonJS(web_component_exports);

// src/html-serializer.ts
var DEFAULT_OPTIONS = {
  classPrefix: "cm-",
  theme: "github",
  linkTarget: "_blank",
  sanitize: false,
  plugins: [],
  customRenderers: {}
};
function resolveOptions(options) {
  return {
    ...DEFAULT_OPTIONS,
    ...options,
    plugins: options?.plugins ?? [],
    customRenderers: options?.customRenderers ?? {}
  };
}
function applySanitize(html, sanitize) {
  if (typeof sanitize === "function") {
    return sanitize(html);
  }
  return html;
}
function blocksToHTML(blocks, options) {
  const opts = resolveOptions(options);
  const prefix = opts.classPrefix;
  let transformedBlocks = blocks;
  for (const plugin of opts.plugins) {
    if (plugin.transformBlock) {
      transformedBlocks = transformedBlocks.map(plugin.transformBlock);
    }
  }
  const htmlParts = transformedBlocks.map((block) => {
    for (const plugin of opts.plugins) {
      if (plugin.renderBlock) {
        const result = plugin.renderBlock(block, () => renderBlock(block, opts));
        if (result !== null) {
          return result;
        }
      }
    }
    const customRenderer = opts.customRenderers[block.type];
    if (customRenderer) {
      return customRenderer(block);
    }
    return renderBlock(block, opts);
  });
  let html = `<div class="${prefix}preview">${htmlParts.join("\n")}</div>`;
  html = applySanitize(html, opts.sanitize);
  return html;
}
async function renderAsync(blocks, options) {
  const opts = resolveOptions(options);
  for (const plugin of opts.plugins) {
    if (plugin.init) {
      await plugin.init();
    }
  }
  let html = blocksToHTML(blocks, options);
  for (const plugin of opts.plugins) {
    if (plugin.postProcess) {
      html = await plugin.postProcess(html);
    }
  }
  return html;
}
function renderBlock(block, opts) {
  const prefix = opts.classPrefix;
  switch (block.type) {
    case "paragraph":
      return `<p class="${prefix}paragraph">${renderInlineContent(block.content, opts)}</p>`;
    case "heading":
      return renderHeading(block, opts);
    case "bulletList":
      return renderBulletList(block, opts);
    case "numberedList":
      return renderNumberedList(block, opts);
    case "checkList":
      return renderCheckList(block, opts);
    case "codeBlock":
      return renderCodeBlock(block, opts);
    case "blockquote":
      return `<blockquote class="${prefix}blockquote">${renderInlineContent(block.content, opts)}</blockquote>`;
    case "table":
      return renderTable(block, opts);
    case "image":
      return renderImage(block, opts);
    case "divider":
      return `<hr class="${prefix}divider" />`;
    case "callout":
      return renderCallout(block, opts);
    default:
      return `<div class="${prefix}unknown">${renderInlineContent(block.content, opts)}</div>`;
  }
}
function renderHeading(block, opts) {
  const prefix = opts.classPrefix;
  const level = block.props.level;
  const tag = `h${level}`;
  const content = renderInlineContent(block.content, opts);
  return `<${tag} class="${prefix}heading ${prefix}h${level}">${content}</${tag}>`;
}
function renderBulletList(block, opts) {
  const prefix = opts.classPrefix;
  const items = block.children.map((child) => `<li>${renderInlineContent(child.content, opts)}</li>`).join("\n");
  return `<ul class="${prefix}bullet-list">
${items}
</ul>`;
}
function renderNumberedList(block, opts) {
  const prefix = opts.classPrefix;
  const items = block.children.map((child) => `<li>${renderInlineContent(child.content, opts)}</li>`).join("\n");
  return `<ol class="${prefix}numbered-list">
${items}
</ol>`;
}
function renderCheckList(block, opts) {
  const prefix = opts.classPrefix;
  const checked = block.props.checked;
  const checkboxAttr = checked ? "checked disabled" : "disabled";
  const textClass = checked ? `${prefix}checked` : "";
  return `
<div class="${prefix}checklist-item">
  <input type="checkbox" ${checkboxAttr} />
  <span class="${textClass}">${renderInlineContent(block.content, opts)}</span>
</div>`.trim();
}
function renderCodeBlock(block, opts) {
  const prefix = opts.classPrefix;
  const code = block.content.map((span) => span.text).join("");
  const language = block.props.language || "";
  const escapedCode = escapeHtml(code);
  const langClass = language ? ` language-${language}` : "";
  const dataLang = language ? ` data-language="${language}"` : "";
  return `<pre class="${prefix}code-block"${dataLang}><code class="${prefix}code${langClass}">${escapedCode}</code></pre>`;
}
function renderTable(block, opts) {
  const prefix = opts.classPrefix;
  const { headers, rows, alignments } = block.props;
  const getStyle = (index) => {
    const align = alignments?.[index];
    return align ? ` style="text-align: ${align}"` : "";
  };
  const headerRow = headers.length > 0 ? `<thead><tr>${headers.map((h, i) => `<th${getStyle(i)}>${escapeHtml(h)}</th>`).join("")}</tr></thead>` : "";
  const bodyRows = rows.map((row) => `<tr>${row.map((cell, i) => `<td${getStyle(i)}>${escapeHtml(cell)}</td>`).join("")}</tr>`).join("\n");
  return `<table class="${prefix}table">
${headerRow}
<tbody>
${bodyRows}
</tbody>
</table>`;
}
function renderImage(block, opts) {
  const prefix = opts.classPrefix;
  const { url, alt, title, width, height } = block.props;
  const altAttr = alt ? ` alt="${escapeHtml(alt)}"` : ' alt=""';
  const titleAttr = title ? ` title="${escapeHtml(title)}"` : "";
  const widthAttr = width ? ` width="${width}"` : "";
  const heightAttr = height ? ` height="${height}"` : "";
  const img = `<img src="${escapeHtml(url)}"${altAttr}${titleAttr}${widthAttr}${heightAttr} />`;
  const caption = alt ? `<figcaption>${escapeHtml(alt)}</figcaption>` : "";
  return `<figure class="${prefix}image">${img}${caption}</figure>`;
}
function renderCallout(block, opts) {
  const prefix = opts.classPrefix;
  const type = block.props.type;
  const content = renderInlineContent(block.content, opts);
  return `
<div class="${prefix}callout ${prefix}callout-${type}" role="alert">
  <strong class="${prefix}callout-title">${type}</strong>
  <div class="${prefix}callout-content">${content}</div>
</div>`.trim();
}
function renderInlineContent(spans, opts) {
  return spans.map((span) => renderSpan(span, opts)).join("");
}
function renderSpan(span, opts) {
  let html = escapeHtml(span.text);
  const styles = span.styles;
  if (styles.code) {
    html = `<code>${html}</code>`;
  }
  if (styles.highlight) {
    html = `<mark>${html}</mark>`;
  }
  if (styles.strikethrough) {
    html = `<del>${html}</del>`;
  }
  if (styles.underline) {
    html = `<u>${html}</u>`;
  }
  if (styles.italic) {
    html = `<em>${html}</em>`;
  }
  if (styles.bold) {
    html = `<strong>${html}</strong>`;
  }
  if (styles.link) {
    const target = opts.linkTarget === "_blank" ? ' target="_blank" rel="noopener noreferrer"' : "";
    const title = styles.link.title ? ` title="${escapeHtml(styles.link.title)}"` : "";
    html = `<a href="${escapeHtml(styles.link.url)}"${title}${target}>${html}</a>`;
  }
  return html;
}
function escapeHtml(text) {
  return text.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;");
}

// src/web-component.ts
async function lazyParse(markdown) {
  try {
    const core = await import("@create-markdown/core");
    return core.parse(markdown);
  } catch {
    throw new Error(
      "@create-markdown/core is required to parse markdown in <markdown-preview>. Install it, or provide pre-parsed blocks via the blocks attribute / setBlocks()."
    );
  }
}
var MarkdownPreviewElement = class extends HTMLElement {
  constructor() {
    super();
    __publicField(this, "_shadow", null);
    __publicField(this, "plugins", []);
    __publicField(this, "defaultTheme", "github");
    __publicField(this, "styleElement");
    __publicField(this, "contentElement");
    const mode = this.constructor._shadowMode;
    if (mode !== "none") {
      this._shadow = this.attachShadow({ mode });
    }
    this.styleElement = document.createElement("style");
    this.renderRoot.appendChild(this.styleElement);
    this.contentElement = document.createElement("div");
    this.contentElement.className = "markdown-preview-content";
    this.renderRoot.appendChild(this.contentElement);
    this.updateStyles();
  }
  static get observedAttributes() {
    return ["theme", "link-target", "async"];
  }
  get renderRoot() {
    return this._shadow ?? this;
  }
  connectedCallback() {
    this.render();
  }
  attributeChangedCallback(_name, _oldValue, _newValue) {
    this.render();
  }
  setPlugins(plugins) {
    this.plugins = plugins;
    this.render();
  }
  setDefaultTheme(theme) {
    this.defaultTheme = theme;
    this.render();
  }
  getMarkdown() {
    const blocksAttr = this.getAttribute("blocks");
    if (blocksAttr) {
      try {
        const blocks = JSON.parse(blocksAttr);
        return blocks.map((b) => b.content.map((s) => s.text).join("")).join("\n\n");
      } catch {
        return "";
      }
    }
    return this.textContent || "";
  }
  setMarkdown(markdown) {
    this.textContent = markdown;
    this.render();
  }
  setBlocks(blocks) {
    this.setAttribute("blocks", JSON.stringify(blocks));
    this.render();
  }
  getOptions() {
    const theme = this.getAttribute("theme") || this.defaultTheme;
    const linkTarget = this.getAttribute("link-target") || "_blank";
    return {
      theme,
      linkTarget,
      plugins: this.plugins
    };
  }
  async getBlocks() {
    const blocksAttr = this.getAttribute("blocks");
    if (blocksAttr) {
      try {
        return JSON.parse(blocksAttr);
      } catch {
        console.warn("Invalid blocks JSON in markdown-preview element");
        return [];
      }
    }
    const markdown = this.textContent || "";
    return lazyParse(markdown);
  }
  async render() {
    const blocks = await this.getBlocks();
    const options = this.getOptions();
    const isAsync = this.hasAttribute("async") || this.plugins.length > 0;
    try {
      let html;
      if (isAsync) {
        html = await renderAsync(blocks, options);
      } else {
        html = blocksToHTML(blocks, options);
      }
      this.contentElement.innerHTML = html;
    } catch (error) {
      console.error("Error rendering markdown preview:", error);
      this.contentElement.innerHTML = `<div class="error">Error rendering content</div>`;
    }
  }
  updateStyles() {
    const pluginCSS = this.plugins.filter((p) => p.getCSS).map((p) => p.getCSS()).join("\n\n");
    const hostRule = this._shadow ? ":host { display: block; }" : "markdown-preview { display: block; }";
    this.styleElement.textContent = `
${hostRule}

.markdown-preview-content {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Noto Sans', Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 1.6;
}

.error {
  color: #cf222e;
  padding: 1rem;
  background: #ffebe9;
  border-radius: 6px;
}

${pluginCSS}
    `.trim();
  }
};
/** @internal */
__publicField(MarkdownPreviewElement, "_shadowMode", "open");
var globalPlugins = [];
var globalDefaultTheme = "github";
function registerPreviewElement(options) {
  const tagName = options?.tagName || "markdown-preview";
  const plugins = options?.plugins || [];
  const defaultTheme = options?.defaultTheme || "github";
  const shadowMode = options?.shadowMode ?? "open";
  globalPlugins = plugins;
  globalDefaultTheme = defaultTheme;
  if (!customElements.get(tagName)) {
    class ConfiguredMarkdownPreview extends MarkdownPreviewElement {
      constructor() {
        super();
        this.setPlugins(globalPlugins);
        this.setDefaultTheme(globalDefaultTheme);
      }
    }
    __publicField(ConfiguredMarkdownPreview, "_shadowMode", shadowMode);
    customElements.define(tagName, ConfiguredMarkdownPreview);
  }
}
function autoRegister() {
  if (typeof window !== "undefined" && typeof customElements !== "undefined") {
    registerPreviewElement();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MarkdownPreviewElement,
  autoRegister,
  registerPreviewElement
});
